mux/src/tmux_pty.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    tmux::{RefTmuxRemotePane, TmuxCmdQueue, TmuxDomainState},
+    tmux_commands::SendKeys,
+};
+use portable_pty::{Child, ExitStatus, MasterPty};
+use std::{
+    io::{Read, Write},
+    sync::{Arc, Condvar, Mutex},
+};
+
+pub(crate) struct TmuxReader {
+    rx: flume::Receiver<String>,
+    // If a string received from rx is larger then the buffer to write out,
+    // we put that string here and use a cursor to indicate all chars before
+    // that cursor have been written out.
+    // Clear this buffer before receive next string.
+    head_buffer: String,
+    head_cursor: usize, // the first char of next write
+}
+
+impl Read for TmuxReader {
+    fn read(&mut self, mut buf: &mut [u8]) -> std::io::Result<usize> {
+        if !self.head_buffer.is_empty() {
+            let mut buffer_cleared = false;
+            let bytes = if self.head_cursor + buf.len() >= self.head_buffer.len() {
+                buffer_cleared = true;
+                &self.head_buffer[self.head_cursor..]
+            } else {
+                &self.head_buffer[self.head_cursor..(self.head_cursor + buf.len())]
+            };
+            return buf.write(bytes.as_bytes()).map(|res| {
+                // update buffer if write success
+                if buffer_cleared {
+                    self.head_buffer.clear();
+                    self.head_cursor = 0;
+                } else {
+                    self.head_cursor = self.head_cursor + buf.len();
+                }
+                res
+            });
+        } else {
+            match self.rx.recv() {
+                Ok(str) => {
+                    if str.len() > buf.len() {
+                        self.head_buffer = str;
+                        self.head_cursor = 0;
+                        return self.read(buf);
+                    } else {
+                        return buf.write(str.as_bytes());
+                    }
+                }
+                Err(_) => {
+                    return Ok(0);
+                }
+            }
```

Comment:
Add a buffer here to prevent write out of range. Will this solve the problem?

Reply:
I haven't looked deeply at this change yet, but my knee-jerk reaction is that that looks very complex.  I think replacing the channel based reader/writer with a socketpair will make most of the complexity disappear!